### PR TITLE
front: refactor refs allocation + enforce guardrails on actions per step

### PR DESCRIPTION
**Pull request by spolu**
## Description

Fixes: https://github.com/dust-tt/tasks/issues/919
(context in the eng runner issue)

This PR:
- Cap the total number of actions per step to 16
- Share topK among actions associated with the same configuration in the same step (we've seen examples of assistants generating 22 retrievals. This would share the topK among them)
- Finer-grained allocations of Refs. Each action can see the other actions of the current step and can compute their topK and as a result the refs they consume.

cc @PopDaph related to websearch.

Let's test this in production and we'll need a bit more refactor to move to websearch such that each action can expose its computation of references usage (we just need to add a new function on `BaseActionConfigurationServerRunner` and a shared tooling that will do what we do here across retrieval and websearch.

## Risk

N/A

## Deploy Plan

- deploy `front`

### CODE REVIEW

File: front/lib/api/assistant/actions/retrieval.ts [hunk 1]
```diff
@@ -546,7 +607,17 @@ export class RetrievalConfigurationServerRunner extends BaseActionConfigurationS
             token_count: number;
           }[];
 
-          const refs = getRefs().slice(refsOffset, refsOffset + v.length);
+          if (refsOffset + topK > getRefs().length) {
+            // This is a stream dropping error since the guardrails put in place should prevent this
+            // from ever happeaning (max 16 actions per step and sharing of topK among actions of
+            // the same type).
+            throw new Error(
```
Comment (fontanierh) [inline front/lib/api/assistant/actions/retrieval.ts:614]: let's also panic ? if this starts happening we need to know ASAP

File: front/lib/api/assistant/agent.ts [hunk 2]
```diff
@@ -48,6 +48,7 @@ import { redisClient } from "@app/lib/redis";
 import logger from "@app/logger/logger";
 
 const CANCELLATION_CHECK_INTERVAL = 500;
+const MAX_ACTIONS_PER_STEP = 16;
```
Comment (fontanierh) [inline front/lib/api/assistant/agent.ts:51]: even 16 is quite a lot, maybe 8 would be a better default ?

Comment (fontanierh) [reply front/lib/api/assistant/agent.ts:51]: def not strongly held though

**Review by fontanierh — APPROVED** [review]
🙏

File: front/lib/api/assistant/actions/retrieval.ts (hunk 1 above)
Comment (spolu) [reply front/lib/api/assistant/actions/retrieval.ts:614]: This triggers a stream dropping error which is ~~~ the highest level of panic we have today

File: front/lib/api/assistant/agent.ts (hunk 2 above)
Comment (spolu) [reply front/lib/api/assistant/agent.ts:51]: Since topK are shared I think 16 can be useful to get some breadth int he results

**Pull request merged**